src/NuGet.Core/NuGet.Packaging/Signing/Authoring/Signer.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Signing
+{
+    /// <summary>
+    /// Remove or add signature package metadata.
+    /// </summary>
+    public sealed class Signer
+    {
+        private readonly ISignedPackage _package;
+        private readonly SigningSpecificationsV1 _specifications = SigningSpecifications.V1;
+        private readonly ISignatureProvider _signatureProvider;
+
+        /// <summary>
+        /// Creates a signer for a specific package.
+        /// </summary>
+        /// <param name="package">Package to sign or modify.</param>
+        public Signer(ISignedPackage package)
+            : this(package, new X509SignatureProvider(new TimestampProvider()))
+        {
+        }
+
+        /// <summary>
+        /// Creates a signer for a specific package.
+        /// </summary>
+        /// <param name="package">Package to sign or modify.</param>
+        public Signer(ISignedPackage package, ISignatureProvider signatureProvider)
+        {
+            _package = package ?? throw new ArgumentNullException(nameof(package));
+            _signatureProvider = signatureProvider ?? throw new ArgumentNullException(nameof(signatureProvider));
+        }
+
+        /// <summary>
+        /// Add a signature to a package.
+        /// </summary>
+        public async Task SignAsync(SignPackageRequest request, ILogger logger, CancellationToken token)
+        {
+            // Verify hash is allowed
+
+            // Generate manifest
+            var packageEntries = await _package.GetContentManifestEntriesAsync(request.HashAlgorithm, token);
+
+            var manifest = new PackageContentManifest(
+                PackageContentManifest.DefaultVersion,
+                request.HashAlgorithm,
+                packageEntries);
+
+            // Generate manifest hash and add file
+            var manifestHash = await AddManifestAndGetHashAsync(manifest, request.HashAlgorithm, token);
+
+            // Create signature
+            var sig = await _signatureProvider.CreateSignatureAsync(request, manifestHash, logger, token);
```

Comment:
nit: I always prefer having full variable names that are clear about what they are, so `signature` instead of `sig`

Reply:
I'll update it